docs/fn-2.0/README.md

```diff
@@ -0,0 +1,416 @@
+* Functions are packaged in container images
+* Input is a stream of CloudEvent messages in JSON format
+* Output is a stream of CloudEvent messages in JSON format
+  * With some additions to deal with things like error handling and protocol specific responses (eg: function wants to return a specific http status code if it was fired from an HTTP event source.)
+
+#### Function Error Handling
+
+Errors will be returned as a CloudEvent with an “error” attribute, no “data” attribute, describing the error. 
+
+```
+{
+  "cloud-events-version": "0.1",
+  "event-id": "6480da1a-5028-4301-acc3-fbae628207b3",
+  "source": "http://example.com/repomanager",
+  "event-type": "http-request",
+  "event-type-version": "v1.5",
+  "event-time": "2018-04-01T23:12:34Z",
+  "content-type": "application/json",
+  “error”: {
+    "message": "an error occurred, blah, blah",
+    "type": "Exception",
+    "trace": "stack trace"
+  },
+  “extensions”: {
+    “protocol”: {
+       “headers”: …
+       Etc, same as we do now for JSON format. 
+    }
+  }
+}
+```
+
+Error map:
+* message - required
+* type - type of error, eg IOError - optional
+* trace - stack trace - optional
+
+### Triggers
+
+Triggers bind events to functions along with configuration.
+
+Eg: HTTP event
+
+```json
+{
+    "source": "http://myapp.com/x"
+}
+```
+
+Trigger will look up the source in the triggers db then update the event with the Function information plus the app and trigger configuration.
+
+```json
+{
+    "extensions": {
+        "trigger": {
+            "function_url": "func url",
+            "config": {
+                MAP OF CONFIGS FROM APP AND TRIGGER
+            }
+        }
+    }
+}
+```
+
+It will then pass this message along to the Fn Router/LB for execution.
+
+### Apps/Applications
+
+Applications are a grouping of triggers to make up a complete application. For example, an application might have the following triggers:
+
+* http: /x -> trigger maps /x -> function f1 with config c1
+* http: /y …
+* schedule: run function f2 every day at 8am
+* queue: listen to topic t1 filtering on ff1, for every event run function f3
+
+All of these things make up the application. Applications also define 
+additional annotations and configuration to be added to each trigger.
+
+## Components
+
+The following components make up the Fn platform. 
+
+* All components can run separately
+* They all have their own main
+* One main main to tie them all into one bin for a nice developer experience.
+
+### Runner
+
+The Runner does nothing more than take an event message, execute the function and return results.
+
+* Main endpoint `/run` with CloudEvent as input.
+* When called, runs a function and returns the results, as a CloudEvent.
+* If busy, returns 503.
+* Event input contains everything it needs to run, including config/secrets.
+  * TODO: Need to be sure we don’t log any secrets
+* Logs can be retrieved separately via `/log` endpoint so main response can be returned faster.
+
```

Comment:
As regards Logs, do you think that revisiting the logging is in scope for the new version or not? Maybe I am wrong but the logging experience at the moment is not the best one from a customer point of view and from an operator. I think we can't build a system which has a fully completed logging solution but there were some interesting proposals (https://github.com/fnproject/fn/issues/337) for adding hooks so as it will be possible to offload logging to an external backend and some other open questions about the format of the log (https://github.com/fnproject/fn/issues/536).

Reply:
IMO, I don't think we can have a service that doesn't have logs. Nobody would use it.

Reply:
one thought was that we could root logs on `/apps/:app/logs`, which could be more useful to users to get a 'stream' of logs across all triggers on their app vs. having to pull all of these out of individual calls (there is a TODO in the swagger for this) -- this is not trivial to implement in s3 afaik (no append). largely, this is unanswered. I'm open to ideas on improvements. From an OSS angle, to a certain degree all we can do is offer an API that is reasonable to implement and has a decent out of the box options (s3/minio), most end users of this will need something more integrated and robust than what we can offer (maybe? we could have a good OSS story around ELK instead of minio too, and maybe that's OK). I'm not sure our current offering is great and open to discussion there; this is a good opportunity to fix it even if it's largely unrelated to the core changes here. part of me thinks that a lot of the answers here are 'better tooling', most logs seem likely to go unread or streamed elsewhere to be analyzed (syslog hooks is in progress, but we need a good way to debug without that).

Reply:
Yeah I agree better tooling. @treeder I was not saying to not have logs. I was saying to by default not store them.. And make it simple to change that.. Because logs can really eat up disk space and storing them locally by default in db seems meh to me. Or have an option to not store one of those would be nice.

Either way current way is not that useful IMO